libs/langchain-google-common/src/media.ts

```diff
@@ -0,0 +1,775 @@
+import { getEnvironmentVariable } from "@langchain/core/utils/env";
+import {
+  MediaBlob,
+  BlobStore,
+  BlobStoreOptions,
+} from "./utils/media_core.js";
+import {
+  GoogleConnectionParams,
+  GoogleRawResponse,
+  GoogleResponse,
+} from "./types.js";
+import { GoogleHostConnection, GoogleRawConnection } from "./connection.js";
+import {
+  ApiKeyGoogleAuth,
+  GoogleAbstractedClient,
+  GoogleAbstractedClientOpsMethod,
+} from "./auth.js";
+
+export interface GoogleUploadConnectionParams<AuthOptions>
+  extends GoogleConnectionParams<AuthOptions> {}
+
+export abstract class GoogleMultipartUploadConnection<
+  CallOptions extends AsyncCallerCallOptions,
+  ResponseType extends GoogleResponse,
+  AuthOptions
+> extends GoogleHostConnection<CallOptions, ResponseType, AuthOptions> {
+  constructor(
+    fields: GoogleConnectionParams<AuthOptions> | undefined,
+    caller: AsyncCaller,
+    client: GoogleAbstractedClient
+  ) {
+    super(fields, caller, client);
+  }
+
+  async _body(
+    separator: string,
+    data: MediaBlob,
+    metadata: Record<string, unknown>
+  ): Promise<string> {
+    const contentType = data.mimetype;
+    const { encoded, encoding } = await data.encode();
+    const body = [
+      `--${separator}`,
+      "Content-Type: application/json; charset=UTF-8",
+      "",
+      JSON.stringify(metadata),
+      "",
+      `--${separator}`,
+      `Content-Type: ${contentType}`,
+      `Content-Transfer-Encoding: ${encoding}`,
+      "",
+      encoded,
+      `--${separator}--`,
+    ];
+    return body.join("\n");
+  }
+
+  async request(
+    data: MediaBlob,
+    metadata: Record<string, unknown>,
+    options: CallOptions
+  ): Promise<ResponseType> {
+    const separator = `separator-${Date.now()}`;
+    const body = await this._body(separator, data, metadata);
+    const requestHeaders = {
+      "Content-Type": `multipart/related; boundary=${separator}`,
+      "X-Goog-Upload-Protocol": "multipart",
+    };
+    const response = this._request(body, options, requestHeaders);
+    return response;
+  }
+}
+
+export abstract class GoogleDownloadConnection<
+  CallOptions extends AsyncCallerCallOptions,
+  ResponseType extends GoogleResponse,
+  AuthOptions
+> extends GoogleHostConnection<CallOptions, ResponseType, AuthOptions> {
+  async request(options: CallOptions): Promise<ResponseType> {
+    return this._request(undefined, options);
+  }
+}
+
+export abstract class GoogleDownloadRawConnection<
+  CallOptions extends AsyncCallerCallOptions,
+  AuthOptions
+> extends GoogleRawConnection<CallOptions, AuthOptions> {
+  buildMethod(): GoogleAbstractedClientOpsMethod {
+    return "GET";
+  }
+
+  async request(options: CallOptions): Promise<GoogleRawResponse> {
+    return this._request(undefined, options);
+  }
+}
+
+export interface BlobStoreGoogleParams<AuthOptions>
+  extends GoogleConnectionParams<AuthOptions>,
+    AsyncCallerParams,
+    BlobStoreOptions
+  {}
+
+export abstract class BlobStoreGoogle<
```

Comment:
Do we absolutely need a second layer of inheritance?

Reply:
Would prefer making this the base level for now until something along the lines of `BlobStore`/`MediaManager` makes it into core CC @eyurtsev

Reply:
Currently there are two different BlobStores from Google (and it is arguable there might be a third) that take advantage of the existing `AuthOptions` mechanism that we have in place.

But there are also BlobStores that are unrelated to the Google auth mechanism itself. They provide the `MediaBlob` store/fetch semantics, but don't need the Google auth frameworks.

I'm not opposed to collapsing a level, but I think at the moment that we gain some advantages from the abstraction. Happy to discuss further.